mtenn/combination.py

```diff
@@ -0,0 +1,413 @@
+        grad_dict: dict[str, List[torch.Tensor]]
+            Dict mapping from parameter name to list of gradients
+        param_names: List[str]
+            List of parameter names
+        model_params: torch.Tensor
+            Actual parameters that we'll return the gradients for. Each param
+            should be passed individually for the backward pass to work right.
+        """
+        neg = (-1) ** neg
+        # Calculate once for reuse later
+        all_preds = torch.stack(pred_list).flatten()
+        adj_preds = neg * scale * all_preds.detach()
+        Q = torch.logsumexp(adj_preds, dim=0)
+        # Calculate the actual prediction
+        final_pred = (neg * Q / scale).detach()
+
+        return final_pred
+
+    @staticmethod
+    def setup_context(ctx, inputs, output):
+        neg, scale, pred_list, grad_dict, param_names, *model_params = inputs
+
+        grad_dict_keys, grad_dict_tensors = Combination.split_grad_dict(grad_dict)
+
+        # Save non-Tensors for backward
+        ctx.neg = neg
+        ctx.scale = scale
+        ctx.grad_dict_keys = grad_dict_keys
+        ctx.param_names = param_names
+
+        # Save Tensors for backward
+        # Saving:
+        #  * Predictions (1 tensor)
+        #  * Grad tensors (N params * M poses tensors)
+        #  * Model param tensors (N params tensors)
+        ctx.save_for_backward(
+            torch.stack(pred_list).flatten(),
+            *grad_dict_tensors,
+            *model_params,
+        )
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        # Unpack saved tensors
+        preds, *other_tensors = ctx.saved_tensors
+
+        # Split up other_tensors
+        grad_dict_tensors = other_tensors[: len(ctx.grad_dict_keys)]
+
+        grad_dict = Combination.join_grad_dict(ctx.grad_dict_keys, grad_dict_tensors)
+
+        # Begin calculations
+        neg = (-1) ** ctx.neg
+
+        # Calculate once for reuse later
+        adj_preds = neg * ctx.scale * preds.detach()
+        Q = torch.logsumexp(adj_preds, dim=0)
+
+        # Calculate final gradients for each parameter
+        final_grads = {}
+        for n, grad_list in grad_dict.items():
+            final_grads[n] = (
+                torch.stack(
+                    [
+                        grad * (pred - Q).exp()
+                        for grad, pred in zip(grad_list, adj_preds)
+                    ],
+                    axis=-1,
+                )
+                .detach()
+                .sum(axis=-1)
+            )
+
+        # Adjust gradients by grad_output
+        for grad in final_grads.values():
+            grad *= grad_output
+
+        # Pull out return vals
+        return_vals = [None] * 5 + [final_grads[n] for n in ctx.param_names]
+        return tuple(return_vals)
+
+
+class BoltzmannCombination(Combination):
+    """
+    Combine a list of deltaG predictions according to their Boltzmann weight.
+    Treat energy in implicit kT units.
+    """
+
```

Comment:
You could make the associated `_XXX_CombinationFunc` a class attribute of the `XXCombination class` like this:

```python
class BoltzmannCombination(Combination):
    combination_func = _BoltzmannCombinationFunc

...
    def forward(...):
        return self.combination_func.apply(
           *args, **kwargs, 
        )
```

and then you might be able to refactor `forward` into the base class

But you don't have to, this setup is more explicit.

Reply:
I think I'll leave it as is, just because the `_MaxCombinationFunc` has a slightly different signature than the others, so I'd have to do something slightly hacky there I think

Reply:
yeah no worries, I'm happy with that.